File: backend/src/domain/configuration/domain.rs
#![allow(dead_code)]
use std::sync::Arc;

use crate::{
    datasource::{dbdao::dao::DBDao, scylladao::dao::ScyllaDao, vectordao::dao::VectorDao},
    domain::configuration::schema::{ProviderConfig, SystemConfig},
    infra::registry::Registry,
};

#[derive(Clone)]
pub struct ConfigurationDomain {
    pub db_dao: Arc<DBDao>,
    pub scylla_dao: Arc<ScyllaDao>,
    pub vector_dao: Arc<VectorDao>,
}

impl ConfigurationDomain {
    pub fn new(r: &Registry) -> Self {
        Self {
            db_dao: r.db_dao.clone(),
            scylla_dao: r.scylla_dao.clone(),
            vector_dao: r.vector_dao.clone(),
        }
    }

    pub async fn get_system_config(&self, id: i64) -> Option<SystemConfig> {
        self.db_dao
            .get_system_config(id)
            .await
            .ok()
            .map(|record| SystemConfig {
                id: record.id,
                key: record.key,
                value: record.value,
                description: record.description,
                category: record.category,
                created_at: record.created_at,
                updated_at: None,
            })
    }

    pub async fn search_system_configs(
        &self,
        key: Option<String>,
        category: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<SystemConfig>, i64) {
        match self
            .db_dao
            .search_system_configs(key, category, page, limit)
            .await
        {
            Ok(records) => {
                let configs: Vec<SystemConfig> = records
                    .into_iter()
                    .map(|record| SystemConfig {
                        id: record.id,
                        key: record.key,
                        value: record.value,
                        description: record.description,
                        category: record.category,
                        created_at: record.created_at,
                        updated_at: None,
                    })
                    .collect();
                let total = configs.len() as i64;
                (configs, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_system_config(
        &self,
        key: String,
        value: String,
        description: Option<String>,
        category: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .add_system_config(key.as_str(), value.as_str(), description, category)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_system_config(
        &self,
        id: i64,
        key: Option<String>,
        value: Option<String>,
        description: Option<String>,
        category: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_system_config(id, key, value, description, category)
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn get_provider_config(&self, id: i64) -> Option<ProviderConfig> {
        self.db_dao
            .get_provider_config(id)
            .await
            .ok()
            .map(|record| ProviderConfig {
                id: record.id,
                provider_name: record.provider_name,
                config_key: record.config_key,
                config_value: record.config_value,
                description: record.description,
                created_at: record.created_at,
                updated_at: None,
            })
    }

    pub async fn search_provider_configs(
        &self,
        provider_name: Option<String>,
        config_key: Option<String>,
        page: i64,
        limit: i64,
    ) -> (Vec<ProviderConfig>, i64) {
        match self
            .db_dao
            .search_provider_configs(provider_name, config_key, page, limit)
            .await
        {
            Ok(records) => {
                let configs: Vec<ProviderConfig> = records
                    .into_iter()
                    .map(|record| ProviderConfig {
                        id: record.id,
                        provider_name: record.provider_name,
                        config_key: record.config_key,
                        config_value: record.config_value,
                        description: record.description,
                        created_at: record.created_at,
                        updated_at: None,
                    })
                    .collect();
                let total = configs.len() as i64;
                (configs, total)
            }
            Err(_) => (Vec::new(), 0),
        }
    }

    pub async fn add_provider_config(
        &self,
        provider_name: String,
        config_key: String,
        config_value: String,
        description: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .add_provider_config(
                provider_name.as_str(),
                config_key.as_str(),
                config_value.as_str(),
                description,
            )
            .await
            .map_err(|e| e.to_string())
    }

    pub async fn update_provider_config(
        &self,
        id: i64,
        provider_name: Option<String>,
        config_key: Option<String>,
        config_value: Option<String>,
        description: Option<String>,
    ) -> Result<i64, String> {
        self.db_dao
            .update_provider_config(id, provider_name, config_key, config_value, description)
            .await
            .map_err(|e| e.to_string())
    }
}

File: backend/domain/configuration/domain.go
package configuration

import (
	"backend/datasource/dbdao"
	"context"
	"errors"
)

type ConfigurationDomain struct {
	DB *dbdao.DB
}

func NewConfigurationDomain(db *dbdao.DB) *ConfigurationDomain {
	return &ConfigurationDomain{DB: db}
}

func (d *ConfigurationDomain) GetSystemConfig(ctx context.Context, id int64) (*SystemConfig, error) {
	cfg, err := d.DB.GetSystemConfig(ctx, id)
	if err != nil {
		return nil, err
	}

	return &SystemConfig{
		ID:          cfg.ID,
		Key:         cfg.Key,
		Value:       cfg.Value,
		Description: cfg.Description,
		Category:    cfg.Category,
		CreatedAt:   &cfg.CreatedAt,
		UpdatedAt:   &cfg.UpdatedAt,
	}, nil
}

func (d *ConfigurationDomain) SearchSystemConfigs(
	ctx context.Context,
	key, category string,
	page, limit int64,
) ([]*SystemConfig, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}

	cfgs, total, err := d.DB.SearchSystemConfigs(ctx, key, category, page, limit)
	if err != nil {
		return nil, 0, err
	}

	result := make([]*SystemConfig, len(cfgs))
	for i, c := range cfgs {
		result[i] = &SystemConfig{
			ID:          c.ID,
			Key:         c.Key,
			Value:       c.Value,
			Description: c.Description,
			Category:    c.Category,
			CreatedAt:   &c.CreatedAt,
			UpdatedAt:   &c.UpdatedAt,
		}
	}

	return result, total, nil
}

func (d *ConfigurationDomain) AddSystemConfig(
	ctx context.Context,
	key, value, description, category string,
) (*SystemConfig, error) {
	if key == "" || value == "" {
		return nil, errors.New("missing required fields")
	}

	cfg := &dbdao.SystemConfig{
		Key:         key,
		Value:       value,
		Description: description,
		Category:    category,
	}

	created, err := d.DB.CreateSystemConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}

	return &SystemConfig{
		ID:          created.ID,
		Key:         created.Key,
		Value:       created.Value,
		Description: created.Description,
		Category:    created.Category,
		CreatedAt:   &created.CreatedAt,
		UpdatedAt:   &created.UpdatedAt,
	}, nil
}

func (d *ConfigurationDomain) UpdateSystemConfig(
	ctx context.Context,
	id int64,
	key, value, description, category *string,
) (*SystemConfig, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}

	cfg := &dbdao.SystemConfig{ID: id}

	if key != nil {
		cfg.Key = *key
	}
	if value != nil {
		cfg.Value = *value
	}
	if description != nil {
		cfg.Description = *description
	}
	if category != nil {
		cfg.Category = *category
	}

	updated, err := d.DB.UpdateSystemConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}

	return &SystemConfig{
		ID:          updated.ID,
		Key:         updated.Key,
		Value:       updated.Value,
		Description: updated.Description,
		Category:    updated.Category,
		CreatedAt:   &updated.CreatedAt,
		UpdatedAt:   &updated.UpdatedAt,
	}, nil
}

func (d *ConfigurationDomain) GetProviderConfig(ctx context.Context, id int64) (*ProviderConfig, error) {
	cfg, err := d.DB.GetProviderConfig(ctx, id)
	if err != nil {
		return nil, err
	}

	return &ProviderConfig{
		ID:           cfg.ID,
		ProviderName: cfg.ProviderName,
		ConfigKey:    cfg.ConfigKey,
		ConfigValue:  cfg.ConfigValue,
		Description:  cfg.Description,
		CreatedAt:    &cfg.CreatedAt,
		UpdatedAt:    &cfg.UpdatedAt,
	}, nil
}

func (d *ConfigurationDomain) SearchProviderConfigs(
	ctx context.Context,
	providerName, configKey string,
	page, limit int64,
) ([]*ProviderConfig, int64, error) {
	if page < 1 {
		page = 1
	}
	if limit < 1 || limit > 100 {
		limit = 20
	}

	cfgs, total, err := d.DB.SearchProviderConfigs(ctx, providerName, configKey, page, limit)
	if err != nil {
		return nil, 0, err
	}

	result := make([]*ProviderConfig, len(cfgs))
	for i, c := range cfgs {
		result[i] = &ProviderConfig{
			ID:           c.ID,
			ProviderName: c.ProviderName,
			ConfigKey:    c.ConfigKey,
			ConfigValue:  c.ConfigValue,
			Description:  c.Description,
			CreatedAt:    &c.CreatedAt,
			UpdatedAt:    &c.UpdatedAt,
		}
	}

	return result, total, nil
}

func (d *ConfigurationDomain) AddProviderConfig(
	ctx context.Context,
	providerName, configKey, configValue, description string,
) (*ProviderConfig, error) {
	if providerName == "" || configKey == "" || configValue == "" {
		return nil, errors.New("missing required fields")
	}

	cfg := &dbdao.ProviderConfig{
		ProviderName: providerName,
		ConfigKey:    configKey,
		ConfigValue:  configValue,
		Description:  description,
	}

	created, err := d.DB.CreateProviderConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}

	return &ProviderConfig{
		ID:           created.ID,
		ProviderName: created.ProviderName,
		ConfigKey:    created.ConfigKey,
		ConfigValue:  created.ConfigValue,
		Description:  created.Description,
		CreatedAt:    &created.CreatedAt,
		UpdatedAt:    &created.UpdatedAt,
	}, nil
}

func (d *ConfigurationDomain) UpdateProviderConfig(
	ctx context.Context,
	id int64,
	providerName, configKey, configValue, description *string,
) (*ProviderConfig, error) {
	if id <= 0 {
		return nil, errors.New("invalid id")
	}

	cfg := &dbdao.ProviderConfig{ID: id}

	if providerName != nil {
		cfg.ProviderName = *providerName
	}
	if configKey != nil {
		cfg.ConfigKey = *configKey
	}
	if configValue != nil {
		cfg.ConfigValue = *configValue
	}
	if description != nil {
		cfg.Description = *description
	}

	updated, err := d.DB.UpdateProviderConfig(ctx, cfg)
	if err != nil {
		return nil, err
	}

	return &ProviderConfig{
		ID:           updated.ID,
		ProviderName: updated.ProviderName,
		ConfigKey:    updated.ConfigKey,
		ConfigValue:  updated.ConfigValue,
		Description:  updated.Description,
		CreatedAt:    &updated.CreatedAt,
		UpdatedAt:    &updated.UpdatedAt,
	}, nil
}
